pgx-utils/src/inventory/postgres_enum.rs

```diff
@@ -0,0 +1,120 @@
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::{quote, ToTokens, TokenStreamExt};
+use syn::{
+    parse::{Parse, ParseStream},
+    DeriveInput, Generics, ItemEnum,
+};
+use syn::{punctuated::Punctuated, Ident, Token};
+
+/// A parsed `#[derive(PostgresEnum)]` item.
+///
+/// It should be used with [`syn::parse::Parse`] functions.
+///
+/// Using [`quote::ToTokens`] will output the declaration for a `pgx::datum::inventory::InventoryPostgresEnum`.
+///
+/// ```rust
+/// use syn::{Macro, parse::Parse, parse_quote, parse};
+/// use quote::{quote, ToTokens};
+/// use pgx_utils::inventory::PostgresEnum;
+///
+/// # fn main() -> eyre::Result<()> {
+/// let parsed: PostgresEnum = parse_quote! {
+///     #[derive(PostgresEnum)]
+///     enum Demo {
+///         Example,
+///     }
+/// };
+/// let inventory_tokens = parsed.to_token_stream();
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug, Clone)]
+pub struct PostgresEnum {
+    name: Ident,
+    generics: Generics,
+    variants: Punctuated<syn::Variant, Token![,]>,
+}
+
+impl PostgresEnum {
+    pub fn new(
+        name: Ident,
+        generics: Generics,
+        variants: Punctuated<syn::Variant, Token![,]>,
+    ) -> Self {
+        Self {
+            name,
+            generics,
+            variants,
+        }
+    }
+
+    pub fn from_derive_input(derive_input: DeriveInput) -> Result<Self, syn::Error> {
+        let data_enum = match derive_input.data {
+            syn::Data::Enum(data_enum) => data_enum,
+            syn::Data::Union(_) | syn::Data::Struct(_) => {
+                return Err(syn::Error::new(derive_input.ident.span(), "expected enum"))
+            }
+        };
+        Ok(Self::new(
+            derive_input.ident,
+            derive_input.generics,
+            data_enum.variants,
+        ))
+    }
+}
+
+impl Parse for PostgresEnum {
+    fn parse(input: ParseStream) -> Result<Self, syn::Error> {
+        let parsed: ItemEnum = input.parse()?;
+        Ok(Self::new(parsed.ident, parsed.generics, parsed.variants))
+    }
+}
+
+impl ToTokens for PostgresEnum {
+    fn to_tokens(&self, tokens: &mut TokenStream2) {
+        // It's important we remap all lifetimes we spot to `'static` so they can be used during inventory submission.
+        let name = self.name.clone();
+        let mut static_generics = self.generics.clone();
+        for lifetime in static_generics.lifetimes_mut() {
+            lifetime.lifetime.ident = Ident::new("static", Span::call_site());
+        }
+        let (_impl_generics, ty_generics, _where_clauses) = static_generics.split_for_impl();
+
+        let variants = self.variants.iter();
+        let inventory_fn_name =
+            syn::Ident::new(&format!("__pgx_internals_enum_{}", name), Span::call_site());
+        let pg_finfo_fn_name = syn::Ident::new(
+            &format!("pg_finfo_{}_wrapper", inventory_fn_name),
+            Span::call_site(),
+        );
+        let inv = quote! {
+            #[no_mangle]
```

Comment:
I don't believe this `pg_finfo_fn_name` stuff is necessary anymore?  Isn't this a remnant of some previous exploration?

Reply:
Great catch!!! THank you!!!